### DISCUSSION TITLE
Installer & MultiMC Meta support

### DISCUSSION THREAD

**Discussion by KorewaLidesu** [CleanroomMC/Cleanroom#77] [Feedback]
### Summary

To provide a simple and convenient solution to install Cleanroom Loader.

### Goals

- Installer:
    + Simplify, lightweight and convenient GUI with sufficient options for basic user.
    + Advance options for high-end user or modpack dev to generate custom modpack instance.
- MultiMC Meta:
   + Support both MultiMC and PolyMC/Prism.
   + Install directly on without having to import template instance.
   + Easy to update meta if required.

### Non Goals

None.

### Motivation

- First, Cleanroom buildchain is utilized on MinecraftForge implementation for installer. It is good enough for basic installation, but has been really limiting on which launcher can be installed on (as only for Mojang Vanilla launcher).
- Second, Cleanroom Loader is still on its early stage, it is required us to have a sufficient way to testing with each new build artifact on productive environments to ensure the stability.
    + This lead to the birth of CleanroomMMC repo that aim to provide an importable instance for MMC-variant launcher. While this is a only temporarely solution, it has been proved to not sufficient due to small change on how meta handled on MultiMC and PolyMC/Prism.

Hence the creation of this proposal to provide better way for basic user, modpack dev and Cleanroom dev to play/test new version.

### Description

- Installer:
    + Basic GUI:
    Same as existing 
    New choice to switch version 
    Extra options: `Generate MultiMC instance pack`, `Generate PolyMC/Prism instance pack`.
    + Advance GUI:
    Change which template should be use to generate custom instance pack.
    Allow convertion of exist MMC type instance pack to Cleanroom.
    *More options will be added later*

### References

[CleanroomMMC](https://github.com/CleanroomMC/CleanroomMMC)

### Guidelines

- [X] I have and will continue to adhere to the guidelines

**Comment by KorewaLidesu** [comment]
Base:
- [ ] Minimal implementation running on Java 17.
- [ ] Standard template for packing installer.
- [ ] Offline installer as same as Forge.

Improve pin point:
- [ ] (MMC) Install Fugue as optional.
- [ ] (MMC) Add `-XX:+UseZGC -XX:+ZGenerational` as default flags.

Extras:
- [x] (MMC) Read mainClass from version.json to fill template (geeky_kappa)
- [ ] (MMC) Custom jvm flags (allow adding custom flags like `-Djava.system.class.loader`) (geeky_kappa)

**Comment by kappa-maintainer** [comment]
Any default gc flags may lag someone else's computer, maybe we shouldn't do that
